lib/diagnostics_channel.js

```diff
@@ -0,0 +1,143 @@
+'use strict';
+
+const {
+  ArrayPrototypeIncludes,
+  ArrayPrototypePush,
+  FunctionPrototypeBind,
+  ObjectCreate,
+  ObjectSetPrototypeOf,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+// TODO(qard): should there be a C++ publisher interface?
+class ActivePublisher {
+  subscribe(subscription) {
+    if (typeof subscription === 'function') {
+      ArrayPrototypePush(this._subscribers, subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        const fn = FunctionPrototypeBind(subscription.onMessage, subscription);
+        ArrayPrototypePush(this._subscribers, fn);
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      this._subscribers[i](this.name, data);
+    }
+  }
+}
+
+class Publisher {
+  constructor(name) {
+    this._subscribers = undefined;
+    this.ifActive = undefined;
+    this.name = name;
+  }
+
+  subscribe(subscription) {
+    ObjectSetPrototypeOf(this, ActivePublisher.prototype);
+    this.ifActive = this;
+    this._subscribers = [];
+    this.subscribe(subscription);
+  }
+
+  shouldPublish() {
+    return false;
+  }
+
+  publish() {}
+}
+
+class DiagnosticsChannel {
+  constructor() {
+    this._channels = ObjectCreate(null);
+    this._subscribers = [];
+  }
+
+  publisher(name) {
+    let channel = this._channels[name];
+    if (channel) return channel;
+
+    if (typeof name !== 'string') {
```

Comment:
I think param type checking should be move up.

Reply:
That positioning is intentional. It makes the channel lookup case slower to do the type-checking first.

Reply:
I thought there shouldn't be a lookup during publishing at all as publishers should get their channel instance once and use it?
Anyhow, please add some comment why it is done that way otherwise someone will change it later as it "looks wrong".